File: tests/homepage/test-navigation.py
#!/usr/bin/env python3
"""Banc du menu principal à deux niveaux — « Nos prestations ».

Pourquoi dans un navigateur
---------------------------

Le balisage dirait que `aria-expanded` est présent ; il ne dirait pas qu'il
suit l'état réel du sous-menu. Il dirait que le chevron a une classe ; pas
qu'il pivote. Il dirait que le panneau existe ; pas qu'il tient dans la
fenêtre. Et il ne dirait rien du seul comportement que le clavier exige
vraiment : qu'Échap rende le focus au parent, faute de quoi il retombe sur
`<body>` et la tabulation repart du début de la page.

Ce que le banc refuse de laisser passer
---------------------------------------

1. un parent qui serait un lien — il n'existe aucune page « Nos prestations » ;
2. un `aria-expanded` qui ment sur l'état réel, dans un sens ou dans l'autre ;
3. un sous-menu qui ne se referme pas : second clic, Échap, clic au dehors,
   sortie du focus — les quatre chemins sont vérifiés séparément ;
4. Échap qui ne rend pas le focus au parent ;
5. la flèche bas qui n'entre pas dans le sous-menu ;
6. un panneau qui déborde de la fenêtre à une largeur de bureau ;
7. les deux menus offerts en même temps, ou aucun des deux ;
8. sous le burger : un groupe replié dont les enfants ne seraient pas
   atteignables, ou un sous-niveau sans décalage visible ;
9. un « Espace client » devenu un lien — la destination n'existe pas encore ;
10. le moindre débordement horizontal, menu ouvert ou fermé.

Le halo de focus n'est pas mesuré ici : `:focus-visible` ne se déclenche pas de
façon fiable sur un `focus()` programmatique. C'est `test-navigation.php` qui
garantit la règle, et une vérification au clavier réel qui l'a constatée.

Prérequis : Python 3 et Google Chrome. Sans Chrome, sortie en code 2
(prérequis absent) — jamais un succès silencieux.

    python3 test-navigation.py
"""

import http.server
import json
import os
import re
import shutil
import socket
import socketserver
import subprocess
import sys
import tempfile
import threading

RACINE = os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__))))

CHROMES = [
    "/Applications/Google Chrome.app/Contents/MacOS/Google Chrome",
    "/Applications/Chromium.app/Contents/MacOS/Chromium",
    "google-chrome",
    "google-chrome-stable",
    "chromium",
    "chromium-browser",
]

# Le menu attendu, dans l'ordre. C'est la seule description de référence :
# le banc ne devine pas, il compare.
PREMIER_NIVEAU = ["Accueil", "Nos prestations", "Tarifs", "Espace client", "Contact"]
PRESTATIONS = ["Déclaration préalable", "Permis de construire", "Conception de plans"]

echecs = 0


def check(libelle, condition, detail=""):
    global echecs
    if not condition:
        echecs += 1
    print("%-74s %s" % (libelle, "OK" if condition else "ECHEC"))
    if not condition and detail:
        print("    " + detail)


def trouver_chrome():
    for c in CHROMES:
        if os.path.isfile(c):
            return c
        trouve = shutil.which(c)
        if trouve:
            return trouve
    return None


class Silencieux(http.server.SimpleHTTPRequestHandler):
    def log_message(self, *args):
        pass


def servir(racine):
    handler = lambda *a, **k: Silencieux(*a, directory=racine, **k)
    with socket.socket() as s:
        s.bind(("127.0.0.1", 0))
        port = s.getsockname()[1]
    httpd = socketserver.TCPServer(("127.0.0.1", port), handler)
    httpd.allow_reuse_address = True
    threading.Thread(target=httpd.serve_forever, daemon=True).start()
    return httpd, port


def mesurer(chrome, url):
    """Chrome écrit le DOM puis ne rend pas toujours la main : on lit et on termine."""
    profil = tempfile.mkdtemp(prefix="urbizen-navigation-")
    proc = subprocess.Popen(
        [
            chrome,
            "--headless=new",
            "--disable-gpu",
            "--no-first-run",
            "--no-default-browser-check",
            "--disable-extensions",
            "--hide-scrollbars",
            "--user-data-dir=" + profil,
            "--host-resolver-rules=MAP * ~NOTFOUND , EXCLUDE 127.0.0.1",
            "--virtual-time-budget=120000",
            "--dump-dom",
            url,
        ],
        stdout=subprocess.PIPE,
        stderr=subprocess.DEVNULL,
        text=True,
    )
    try:
        dom, _ = proc.communicate(timeout=150)
    except subprocess.TimeoutExpired:
        proc.kill()
        dom, _ = proc.communicate()
    finally:
        shutil.rmtree(profil, ignore_errors=True)

    bloc = re.search(r'<pre id="resultat">(.*?)</pre>', dom or "", re.S)
    if not bloc:
        return None, "aucune mesure produite (Chrome a-t-il rendu la page ?)"

    contenu = bloc.group(1)
    for brut, clair in (("&quot;", '"'), ("&amp;", "&"), ("&lt;", "<"), ("&gt;", ">")):
        contenu = contenu.replace(brut, clair)

    if contenu.startswith("ERREUR:"):
        return None, "la page de mesure a échoué : " + contenu[7:].strip()
    if not contenu.startswith("JSON:"):
        return None, "mesure incomplète — la page en était à « %s »" % contenu.strip()[:40]
    try:
        return json.loads(contenu[5:]), None
    except json.JSONDecodeError as e:
        return None, "mesure illisible : %s" % e


def main():
    chrome = trouver_chrome()
    if not chrome:
        print("Google Chrome introuvable — ce banc rejoue un menu réel et ne")
        print("peut pas s'exécuter sans moteur de rendu.")
        print("PREREQUIS ABSENT — BANC NON EXECUTE")
        sys.exit(2)

    httpd, port = servir(RACINE)
    try:
        url = "http://127.0.0.1:%d/tests/homepage/navigation.html" % port
        donnees, erreur = mesurer(chrome, url)
    finally:
        httpd.shutdown()

    if donnees is None:
        print("Impossible de mesurer : %s" % erreur)
        print("1 CONTROLE(S) EN ECHEC")
        sys.exit(1)

    m = {x["largeur"]: x for x in donnees["mesures"]}
    check("Les 8 largeurs sont mesurées", len(m) == 8, "obtenu : %s" % sorted(m))
    if len(m) != 8:
        print("\n1 CONTROLE(S) EN ECHEC")
        sys.exit(1)

    bureau = [w for w in sorted(m) if m[w]["desktopVisible"]]
    mobile = [w for w in sorted(m) if m[w]["burgerVisible"]]

    # ----------------------------------------------------- la bascule ---------

    faux = ["%dpx" % w for w in sorted(m) if m[w]["desktopVisible"] == m[w]["burgerVisible"]]
    check("Un seul des deux menus est offert, à chaque largeur", not faux, " | ".join(faux))
    check(
        "La bascule tombe bien entre 1100 et 1101 px",
        m[1100]["burgerVisible"] and m[1101]["desktopVisible"],
        "1100 burger=%s | 1101 desktop=%s" % (m[1100]["burgerVisible"], m[1101]["desktopVisible"]),
    )

    # ----------------------------------------------------- aucun faux lien ----

    fautifs = ["%dpx : %d" % (w, m[w]["faussesAncres"]) for w in sorted(m) if m[w]["faussesAncres"]]
    check("Aucun href=\"#\" dans l'en-tête", not fautifs, " | ".join(fautifs))
    pro = ["%dpx" % w for w in sorted(m) if m[w]["liensPro"]]
    check("Aucun lien vers un espace professionnels", not pro, " | ".join(pro))

    deb = ["%dpx → %d" % (w, m[w]["debordement"]) for w in sorted(m) if m[w]["debordement"] > 0]
    check("Aucun débordement horizontal, menu fermé", not deb, " | ".join(deb))

    # ----------------------------------------------------- menu de bureau -----

    check(
        "Le parent « Nos prestations » est un bouton, pas un lien",
        all(m[w]["parentBalise"] == "BUTTON" and m[w]["parentHref"] is None for w in bureau),
        " | ".join("%dpx %s" % (w, m[w]["parentBalise"]) for w in bureau),
    )
    # L'étiquette « bientôt » est écrite en minuscules et mise en capitales par
    # le CSS : `textContent` restitue la source, pas ce que l'œil lit.
    libelles = lambda w: [
        re.sub(r"\s*bientôt$", "", e["texte"], flags=re.I) for e in m[w]["entrees"]
    ]
    check(
        "Le premier niveau est exactement le menu attendu, dans l'ordre",
        all(libelles(w) == PREMIER_NIVEAU for w in bureau),
        " | ".join("%dpx %s" % (w, libelles(w)) for w in bureau),
    )
    check(
        "Au chargement : sous-menu fermé et aria-expanded=false",
        all(not m[w]["initial"]["ouvert"] and m[w]["initial"]["aria"] == "false" for w in bureau),
    )
    check(
        "Au clic : sous-menu ouvert, aria-expanded=true, 3 prestations",
        all(m[w]["ouvert"]["ouvert"] and m[w]["ouvert"]["aria"] == "true"
            and m[w]["ouvert"]["liens"] == len(PRESTATIONS) for w in bureau),
        " | ".join("%dpx %s" % (w, m[w]["ouvert"]) for w in bureau),
    )
    # rotate(180deg) donne matrix(-1, 0, 0, -1, 0, 0). Une valeur identité
    # signifierait que la règle n'a jamais mordu.
    plats = ["%dpx %s" % (w, m[w]["ouvert"]["chevron"]) for w in bureau
             if "-1" not in m[w]["ouvert"]["chevron"]]
    check("Le chevron pivote à l'ouverture", not plats, " | ".join(plats))
    hors = ["%dpx [%s, %s]" % (w, m[w]["ouvert"]["gauche"], m[w]["ouvert"]["droite"])
            for w in bureau if m[w]["ouvert"]["gauche"] < 0 or m[w]["ouvert"]["droite"] > w]
    check("Le panneau tient entièrement dans la fenêtre", not hors, " | ".join(hors))
    deb2 = ["%dpx → %d" % (w, m[w]["ouvert"]["debordement"]) for w in bureau
            if m[w]["ouvert"]["debordement"] > 0]
    check("Aucun débordement horizontal, sous-menu ouvert", not deb2, " | ".join(deb2))

    # ----------------------------------------------------- les 4 fermetures ---

    for cle, libelle in (
        ("secondClic", "Second clic sur le parent"),
        ("echap", "Touche Échap"),
        ("sortieFocus", "Sortie du focus par tabulation"),
        ("clicDehors", "Clic au dehors"),
    ):
        restes = ["%dpx" % w for w in bureau if m[w][cle]["ouvert"]]
        check("%s : le sous-menu se referme" % libelle, not restes, " | ".join(restes))

    check(
        "Second clic : aria-expanded repasse à false",
        all(m[w]["secondClic"]["aria"] == "false" for w in bureau),
    )
    mauvais = ["%dpx → %s" % (w, m[w]["echap"]["focus"]) for w in bureau
               if m[w]["echap"]["focus"] != "button.nav-parent"]
    check("Échap : le focus revient au parent", not mauvais, " | ".join(mauvais))

    # ----------------------------------------------------- entrée au clavier --

    check(
        "Flèche bas : ouvre le sous-menu et y pose le focus",
        all(m[w]["flecheBas"]["ouvert"] and m[w]["flecheBas"]["premier"] for w in bureau),
        " | ".join("%dpx %s" % (w, m[w]["flecheBas"]) for w in bureau),
    )

    # ----------------------------------------------------- tiroir mobile ------

    check(
        "Le burger ouvre le tiroir et l'annonce (aria-expanded=true)",
        all(m[w]["tiroir"]["ouvert"] and m[w]["tiroir"]["aria"] == "true" for w in mobile),
    )
    check(
        "L'intitulé « Nos prestations » y est visible et n'est pas un lien",
        all(m[w]["tiroir"]["groupeVisible"] and m[w]["tiroir"]["groupeEstUnLien"] is False
            for w in mobile),
    )
    check(
        "Les 3 prestations sont atteignables sans second dépliage",
        all(m[w]["tiroir"]["enfants"] == len(PRESTATIONS)
            and m[w]["tiroir"]["enfantsVisibles"] == len(PRESTATIONS) for w in mobile),
        " | ".join("%dpx %s/%s" % (w, m[w]["tiroir"]["enfantsVisibles"], m[w]["tiroir"]["enfants"])
                   for w in mobile),
    )
    # Sans décalage, rien ne distingue un sous-niveau d'une entrée de premier
    # niveau : la hiérarchie serait dans le balisage et nulle part à l'écran.
    plat = ["%dpx → %s" % (w, m[w]["tiroir"]["decalage"]) for w in mobile
            if not m[w]["tiroir"]["decalage"] or m[w]["tiroir"]["decalage"] < 8]
    check("Le sous-niveau est décalé d'au moins 8 px", not plat, " | ".join(plat))
    check(
        "Espace client : aria-disabled, et aucune destination inventée",
        all(m[w]["tiroir"]["bientotAria"] == "true" and m[w]["tiroir"]["bientotHref"] is None
            for w in mobile),
    )
    deb3 = ["%dpx → %d" % (w, m[w]["tiroir"]["debordement"]) for w in mobile
            if m[w]["tiroir"]["debordement"] > 0]
    check("Aucun débordement horizontal, tiroir ouvert", not deb3, " | ".join(deb3))

    print("")
    if echecs:
        print("%d CONTROLE(S) EN ECHEC" % echecs)
        sys.exit(1)
    print("TOUS LES CONTROLES PASSENT")


if __name__ == "__main__":
    main()

File: tests/homepage/test-navigation.php
<?php
/**
 * Banc du menu principal — contrat de balisage et de style.
 *
 * `test-navigation.py` rejoue le menu dans un moteur de rendu ; celui-ci fige
 * ce qui se lit dans les sources et n'a pas besoin de navigateur : la
 * composition du menu, le fait que le parent ne soit pas un lien, les valeurs
 * de couleur et de graisse réellement écrites, et le rendu du pattern sur une
 * PAGE INTERNE — que le banc de fidélité ne voit jamais, puisqu'il le rend
 * toujours en position d'accueil.
 *
 * L'EN-TÊTE RÉELLEMENT SERVI
 *
 * Le thème contient un menu WordPress hérité — `parts/header.html` et
 * `parts/superposition-de-navigation.html`, quatre entrées chacun — qu'AUCUN
 * gabarit n'appelle. Modifier « le menu » là ne changerait rien à l'écran.
 * Un contrôle ci-dessous vérifie que les douze gabarits passent tous par
 * `header-urbizen`, donc par le pattern : c'est le seul en-tête servi.
 *
 * Hors WordPress : les fonctions employées sont doublées ci-dessous.
 * Aucun accès réseau, aucune base de données.
 */

define( 'ABSPATH', __DIR__ );

$racine = dirname( __DIR__, 2 );
$theme  = $racine . '/wordpress/urbizen-child';

$fail = 0;
function check( $label, $cond, $detail = '' ) {
	global $fail;
	if ( ! $cond ) { $fail++; }
	printf( "%-74s %s\n", $label, $cond ? 'OK' : 'ECHEC' );
	if ( ! $cond && '' !== $detail ) { echo '    ' . $detail . "\n"; }
}

// Le menu attendu, dans l'ordre. Seule description de référence : le banc ne
// devine pas, il compare.
const PREMIER_NIVEAU = array( 'Accueil', 'Nos prestations', 'Comment ça marche', 'Tarifs', 'Espace client', 'Contact' );
const PRESTATIONS    = array(
	'https://urbizen.fr/declarations-prealables/' => 'Déclaration préalable',
	'https://urbizen.fr/permis-de-construire/'    => 'Permis de construire',
	'https://urbizen.fr/conception/'              => 'Conception de plans',
);

// -------------------------------------------------- doublons WordPress ------
// `$GLOBALS` plutôt que des constantes : le pattern est rendu plusieurs fois,
// dans des positions différentes, et chaque rendu doit pouvoir les changer.
$GLOBALS['banc_front']     = true;
$GLOBALS['banc_permalien'] = '';

function is_front_page() { return $GLOBALS['banc_front']; }
function is_singular()   { return ! $GLOBALS['banc_front']; }
function get_permalink() { return $GLOBALS['banc_permalien']; }
function home_url( $c = '/' ) { return 'https://urbizen.fr' . $c; }
function untrailingslashit( $s ) { return rtrim( (string) $s, '/\\' ); }
function esc_url( $u ) { return htmlspecialchars( (string) $u, ENT_QUOTES, 'UTF-8' ); }
function get_theme_file_uri( $c = '' ) {
	return 'https://urbizen.fr/wp-content/themes/urbizen-child/' . ltrim( $c, '/' );
}

/** Rend le pattern dans une position donnée. */
function rendre( $fichier, $front, $permalien = '' ) {
	$GLOBALS['banc_front']     = $front;
	$GLOBALS['banc_permalien'] = $permalien;
	ob_start();
	include $fichier;
	return ob_get_clean();
}

$pattern  = $theme . '/patterns/header-accueil.php';
$accueil  = rendre( $pattern, true );
$maquette = file_get_contents( $racine . '/frontend/homepage/index.html' );
$css      = file_get_contents( $racine . '/frontend/homepage/homepage.css' );

/** Extrait le bloc du menu de bureau d'un document. */
function bloc_nav( $html ) {
	return preg_match( '#<nav class="nav-links".*?</nav>#s', $html, $m ) ? $m[0] : '';
}

// ------------------------------------------- 1 · l'en-tête réellement servi --

$gabarits = glob( $theme . '/templates/*.html' );
check( 'Les 12 gabarits sont trouvés', 12 === count( $gabarits ), count( $gabarits ) . ' trouvé(s)' );

$hors = array();
foreach ( $gabarits as $g ) {
	if ( ! str_contains( file_get_contents( $g ), '"slug":"header-urbizen"' ) ) {
		$hors[] = basename( $g );
	}
}
check( 'Tous les gabarits passent par header-urbizen, donc par le pattern',
	array() === $hors, implode( ', ', $hors ) );

// Le menu Hostinger hérité peut rester dans le dépôt ; ce qui compte est qu'il
// ne soit branché nulle part. Le jour où il le serait, on servirait deux menus.
$morts = array();
foreach ( array( 'header', 'superposition-de-navigation' ) as $part ) {
	foreach ( $gabarits as $g ) {
		if ( preg_match( '/"slug":"' . preg_quote( $part, '/' ) . '"/', file_get_contents( $g ) ) ) {
			$morts[] = basename( $g ) . ' → ' . $part;
		}
	}
}
check( 'Le menu WordPress hérité n\'est branché sur aucun gabarit',
	array() === $morts, implode( ', ', $morts ) );

// ------------------------------------------- 2 · composition du menu ---------

$nav = bloc_nav( $accueil );
check( 'Le bloc du menu de bureau est repéré', '' !== $nav );

preg_match_all( '#<(a|button|span)\b[^>]*>(?:(?!</\1>).)*?</\1>#s', $nav, $tout );
// Les entrées de premier niveau : les enfants directs, dont le libellé propre.
preg_match_all( '#>([^<>]+)<#', preg_replace( '#<div class="nav-sous-menu".*?</div>#s', '', $nav ), $m );
$libelles = array_values( array_filter( array_map( 'trim', $m[1] ) ) );
$libelles = array_values( array_diff( $libelles, array( 'bientôt' ) ) );
check( 'Premier niveau : exactement ' . implode( ' · ', PREMIER_NIVEAU ),
	PREMIER_NIVEAU === $libelles, implode( ' | ', $libelles ) );

check( 'Le parent est un <button>, jamais un lien',
	str_contains( $nav, '<button type="button" class="nav-parent"' )
	&& ! preg_match( '#<a[^>]*class="[^"]*nav-parent#', $nav ) );
check( 'Le parent déclare son état et sa cible (aria-expanded + aria-controls)',
	str_contains( $nav, 'aria-expanded="false"' )
	&& str_contains( $nav, 'aria-controls="sous-menu-prestations"' )
	&& str_contains( $nav, 'id="sous-menu-prestations"' ) );
check( 'Le sous-menu est replié à l\'arrivée (attribut hidden)',
	(bool) preg_match( '#<div class="nav-sous-menu" id="sous-menu-prestations" hidden>#', $nav ) );

$manquantes = array();
foreach ( PRESTATIONS as $url => $texte ) {
	if ( ! str_contains( $accueil, '>' . $texte . '</a>' ) || ! str_contains( $accueil, $url ) ) {
		$manquantes[] = $texte;
	}
}
check( 'Les 3 prestations sont dans le sous-menu, aux URL inchangées',
	array() === $manquantes, implode( ', ', $manquantes ) );

// Le sous-menu regroupe ; il ne doit pas laisser de doublon au premier niveau.
$hors_sous_menu = preg_replace( '#<div class="nav-sous-menu".*?</div>#s', '', $nav );
$doublons = array();
foreach ( PRESTATIONS as $url => $texte ) {
	if ( str_contains( $hors_sous_menu, $texte ) ) { $doublons[] = $texte; }
}
check( 'Aucune prestation ne reste au premier niveau', array() === $doublons,
	implode( ', ', $doublons ) );

// ------------------------------------------- 3 · les promesses non tenues ----

// Délimiteur `~` et non `#` : `#` est ici dans le motif, et l'aurait refermé.
check( 'Aucun href vide ni href="#" dans tout l\'en-tête',
	! preg_match( '~href="#?"~', $accueil ) );
check( 'Espace client : un <span aria-disabled>, sans href, ni au bureau ni en mobile',
	2 === preg_match_all( '#<span class="(?:nav|mmenu)-bientot"[^>]*aria-disabled="true"#', $accueil )
	&& ! preg_match( '#class="(?:nav|mmenu)-bientot"[^>]*href=#', $accueil ) );
check( 'Aucune page « Nos prestations » n\'est inventée',
	! str_contains( $accueil, 'nos-prestations' ) && ! str_contains( $accueil, 'prestations/' ) );
check( 'Aucun lien vers un espace professionnels',
	! str_contains( $accueil, 'espace-professionnel' ) );
// L'entrée « Guides » attend la page index du lot G : la poser maintenant
// créerait un lien mort. Ce contrôle échouera le jour où elle arrivera — c'est
// son rôle : il force à créer la page et l'entrée dans le même mouvement.
check( 'Guides : aucune entrée tant que /guides/ n\'existe pas',
	! str_contains( $accueil, '/guides/' ) );

// ------------------------------------------- 4 · le tiroir mobile ------------

check( 'Le tiroir mobile porte le même menu, groupe compris',
	str_contains( $accueil, '<p class="mmenu-groupe">Nos prestations</p>' )
	&& 3 === substr_count( $accueil, 'class="mmenu-enfant"' ) );
check( 'L\'intitulé du groupe mobile n\'est pas un lien',
	! preg_match( '#<a[^>]*class="[^"]*mmenu-groupe#', $accueil ) );
check( 'Les deux appels à l\'action du tiroir sont conservés',
	str_contains( $accueil, 'js-open-inquiry' ) && str_contains( $accueil, '>Démarrer mon projet</a>' ) );

// ------------------------------------------- 5 · page courante ---------------

check( 'Accueil : « Accueil » est marqué page courante',
	1 <= substr_count( $accueil, '<a href="#top" aria-current="page">Accueil</a>' ) );
check( 'Accueil : rien d\'autre n\'est marqué page courante',
	2 === substr_count( $accueil, 'aria-current="page"' ),
	substr_count( $accueil, 'aria-current="page"' ) . ' occurrence(s) — attendu 2 (bureau + mobile)' );

// C'est ici que le banc de fidélité est aveugle : il rend toujours l'accueil.
$tarifs = rendre( $pattern, false, 'https://urbizen.fr/tarifs/' );
check( 'Page interne : « Tarifs » est marqué, et lui seul',
	str_contains( $tarifs, '"https://urbizen.fr/tarifs/" aria-current="page">Tarifs</a>' )
	&& 2 === substr_count( $tarifs, 'aria-current="page"' ) );
check( 'Page interne : « Accueil » cesse de l\'être et pointe vers l\'accueil',
	str_contains( $tarifs, '<a href="https://urbizen.fr/">Accueil</a>' )
	&& ! str_contains( $tarifs, '#top" aria-current' ) );

$conception = rendre( $pattern, false, 'https://urbizen.fr/conception/' );
// « Comment ça marche » vise une section de l'ACCUEIL. Sans le préfixe, l'ancre
// nue chercherait un #methode sur la page interne, où il n'existe pas : le lien
// ne mènerait nulle part, en silence.
check( 'Accueil : l\'ancre « Comment ça marche » est nue (défilement en page)',
	str_contains( $accueil, '<a href="#methode">Comment ça marche</a>' ) );
check( 'Page interne : la même ancre est préfixée par l\'accueil',
	str_contains( $tarifs, '<a href="https://urbizen.fr/#methode">Comment ça marche</a>' )
	&& ! str_contains( $tarifs, '<a href="#methode">' ) );
// L'ancre doit exister là où elle pointe, sinon le lien est mort.
check( 'La section #methode existe bien dans les gabarits d\'accueil',
	str_contains( file_get_contents( $theme . '/templates/front-page.html' ), 'id="methode"' )
	&& str_contains( $maquette, 'id="methode"' ) );

check( 'Page interne : une prestation ouverte allume aussi son groupe',
	str_contains( $conception, 'class="nav-parent is-actif"' )
	&& str_contains( $conception, '"https://urbizen.fr/conception/" aria-current="page">Conception de plans</a>' ) );
check( 'Le groupe reste éteint quand aucune prestation n\'est ouverte',
	str_contains( $tarifs, 'class="nav-parent"' ) && ! str_contains( $tarifs, 'is-actif' ) );

// ------------------------------------------- 6 · maquette et pattern ---------

check( 'La maquette porte le même menu que le pattern',
	bloc_nav( $maquette ) === bloc_nav( $accueil ),
	'la maquette est la source de vérité du banc de fidélité : les deux doivent bouger ensemble' );

// ------------------------------------------- 7 · lisibilité des intitulés ----

// Les valeurs sont lues dans la feuille, pas devinées : c'est le seul moyen de
// détecter qu'une règle a été remise à son ancienne valeur par mégarde.
preg_match( '#^\.nav-links a \{([^}]*)\}#m', $css, $regle );
$decl = $regle[1] ?? '';
check( 'Intitulés : couleur = var(--u-ink), le bleu nuit de la charte',
	str_contains( $decl, 'color: var(--u-ink)' ) && ! str_contains( $decl, 'var(--u-ink-soft)' ), trim( $decl ) );
check( 'Intitulés : graisse 600', str_contains( $decl, 'font-weight: 600' ), trim( $decl ) );
check( 'Intitulés : taille inchangée à 14 px', str_contains( $decl, 'font-size: 14px' ), trim( $decl ) );
check( 'Le parent du groupe suit les mêmes valeurs',
	(bool) preg_match( '#\.nav-parent \{[^}]*font-size: 14px;[^}]*font-weight: 600;[^}]*color: var\(--u-ink\)#s', $css ) );
check( 'Le tiroir mobile suit aussi',
	(bool) preg_match( '#\.mmenu a \{[^}]*color: var\(--u-ink\);[^}]*font-weight: 600#', $css ) );
// Mesuré sur le fond de l'en-tête (#FBFCFD) : --u-brand donne 4,25:1 et
// --u-brand-dk 6,12:1. Avec des libellés à 15,33:1, le premier ferait du
// survol un affaiblissement.
check( 'Survol : vert foncé de la charte, celui qui reste lisible',
	(bool) preg_match( '#\.nav-links a:hover[^{]*\{[^}]*color: var\(--u-brand-dk\)#s', $css )
	&& ! preg_match( '#\.nav-links a:hover[^{]*\{[^}]*color: var\(--u-brand\)[;\s]#s', $css ) );
check( 'Survol : marqué aussi par un soulignement, indépendant de la couleur',
	(bool) preg_match( '#\.nav-links a:hover[^{]*\{[^}]*text-decoration: underline#s', $css ) );
check( 'Page courante : elle garde le bleu nuit, pour ne pas se lire comme un survol',
	(bool) preg_match( '#\[aria-current="page"\][^{]*\{[^}]*color: var\(--u-ink\);[^}]*font-weight: 700#s', $css ) );
check( 'Page courante : signalée par une couleur ET un soulignement',
	(bool) preg_match( '#\[aria-current="page"\][^{]*\{[^}]*text-decoration: underline#s', $css ) );

// Une bordure aurait ajouté 2 px de hauteur à l'entrée courante : dans une
// barre à hauteur fixe et contenu centré, la ligne du menu se serait décalée.
check( 'Page courante : aucun décalage vertical (aucune bordure ajoutée)',
	! preg_match( '#\[aria-current="page"\][^{]*\{[^}]*border-(bottom|top)#s', $css ) );

check( 'Le focus clavier reste visible, et seulement au clavier',
	str_contains( $css, '.nav-parent:focus-visible' )
	&& str_contains( $css, '.nav-sous-menu a:focus-visible' )
	&& ! preg_match( '#\.nav-links a:focus \{#', $css ) );

// ------------------------------------------- 8 · ce qui ne doit pas bouger ---

check( 'Le logo est inchangé : aucune dimension, chargement immédiat',
	! preg_match( '/<img[^>]*logo-urbizen\.png[^>]*(width|height)=/', $accueil )
	&& str_contains( $accueil, 'loading="eager"' ) );
check( 'Le CTA « Démarrer mon projet » est inchangé',
	str_contains( $accueil, 'class="btn btn-primary btn-sm js-start"' )
	&& str_contains( $accueil, 'class="nav-cta-long">Démarrer mon projet</span>' )
	&& str_contains( $accueil, 'class="nav-cta-short"' ) );
check( 'Les icônes téléphone et compte, et le burger, sont inchangés',
	str_contains( $accueil, 'class="icon-btn link-tel"' )
	&& str_contains( $accueil, 'class="icon-btn link-login"' )
	&& str_contains( $accueil, 'class="burger"' ) );
check( 'Aucune règle n\'a été ajoutée sur le CTA ni sur le logo',
	! preg_match( '#\.nav-links[^{]*\.js-start#', $css )
	&& ! preg_match( '#\.nav-links[^{]*\.logo#', $css ) );

echo "\n";
if ( $fail ) {
	echo $fail . " CONTROLE(S) EN ECHEC\n";
	exit( 1 );
}
echo "TOUS LES CONTROLES PASSENT\n";
